Clean version of new section on adding new species (#1397)
popsim-consortium/stdpopsim: docs/development.rst

`Documentation and tutorial curators` help maintain the documentation and tutorials.
This can be as easy as pointing out confusing bits of the documentation in a

> GitHub `issue <http://github.com/popgensims/stdpopsim/issues>`__., or adding or editing

Author: I changed all external links in the docs to the `...`__ format with double underscore. This is apparently more robust to changes in the link text.

ILAN: I NEED TO FIX AND ADJUST THIS !!!

We start by
defining auxiliary objects (``_recombination_rate_data`` and ``_ploidy``)
that specify the recombination rate and ploidy
for each chromosome.
Note that the mitochondrial and plastid genomes are associated
with ploidy of 1 and recombination rate of 0,
and all other chromosomes are associated with a ploidy of 2 and the
genome-wide average recombination rate.

.. code-block:: python

    _mean_recombination_rate = 200 / 124000 / 2 / 1e6
    _recombination_rate_data = {str(j): _mean_recombination_rate for j in range(1, 6)}
    _recombination_rate_data["Mt"] = 0
    _recombination_rate_data["Pt"] = 0

    _species_ploidy = 2
    _ploidy = {
      "1": _species_ploidy,
      "2": _species_ploidy,
      "3": _species_ploidy,
      "4": _species_ploidy,
      "5": _species_ploidy,
      "6": _species_ploidy,
      "Mt": 1,
      "Pt": 1,
    }

We then use another object (``_chromosomes``) to wrap up these
values together with the mutation rate, which is set to the
genome-wide average value for every chromosome.
Chromosome ids and lengths are taken from the ``data`` object defined in
the ``genome_data.py`` file.

.. code-block:: python

> 

Author: I wonder if we should also mention here how to specify bacterial species (with "bacterial recombination"). Gene conversion could/should also be mentioned here. Maybe I should add another header "Recombination models"?

including the **effective population size** and **average generation time**,
along with relevant citations.
For example, the ``_species`` object for *A. thaliana* is specified as follows:

.. code-block:: python

    _species = stdpopsim.Species(
        id="AraTha",
        ensembl_id="arabidopsis_thaliana",
        name="Arabidopsis thaliana",
        common_name="A. thaliana",
        genome=_genome,
        generation_time=1.0,
        population_size=10**4,
        ploidy=_species_ploidy,
        citations=[
            stdpopsim.Citation(
                doi="https://doi.org/10.1890/0012-9658(2002)083[1006:GTINSO]2.0.CO;2",
                year=2002,
                author="Donohue",
                reasons={stdpopsim.CiteReason.GEN_TIME},
            ),
            stdpopsim.Citation(
                doi="https://doi.org/10.1016/j.cell.2016.05.063",
                year=2016,
                author="1001GenomesConsortium",
                reasons={stdpopsim.CiteReason.POP_SIZE},
            ),
        ],
    )

Once these two objects (``_genome`` and ``_species``) are specified,
you should be able to load and simulate the newly added species using ``stdpopsim``.

> 

Author: Should the original contributor also generate tables for the docs, as done when a new demographic model is added?

Reviewer: I don't think there's any tables that aren't autogenerated for species; am I forgetting something?

which they think can be helpful for methods development.
Note that generic models should also be accompanied with a clear citation indicating
their usefulness.

---------------------------------------------
Getting set up to add a new demographic model
---------------------------------------------

If this is your first time implementing a demographic model in ``stdpopsim``, it's a good
idea to take some time browsing the :ref:`sec_catalog`
to see how existing demographic models are coded and documented.
If you have any questions or confusion about formatting or implementing demographic models, please
don't hesitate to `open a new issue <http://github.com/popgensims/stdpopsim/issues>`_.
We're more than happy to answer any questions and help get you up and running.
Before you add any code, be sure to have forked the ``stdpopsim`` repository
and cloned it locally, following the instructions in the `GitHub Workflow`_ section.

All code for a species' demographic models is written in the ``demographic_models.py``
file in that species directory ``stdpopsim/catalog/<SPECIES-ID>/``
(where ``<SPECIES-ID>`` is the six-character identifier of the species;
e.g., CanFam).

> If the species does not currently have any demographic model,

Author: We're not mentioning where code for generic demographic models should be added. I actually tried to look for this in the catalog and couldn't find it. Where is it?

Author: I think we do. See [this part](https://popsim-consortium.github.io/stdpopsim-docs/latest/catalog.html#generic-models) in the catalog docs.

Author: Okay. I found the two current generic models in ``stdpopsim/models.py``. I guess I can say that generic models are defined as Python classes in this file, and simply point to the two examples that are already implemented there.

Author: I pushed a new [commit](https://github.com/popsim-consortium/stdpopsim/pull/1397/commits/7a3afa59fa5b3ff0feb020a3aa1452a4496e4306) with some text about this, all at the bottom of the **"What models are appropriate to add?"** section.

Reviewer 2: I'm still unsure whether we actually want people to contribute generic models with no associated species. As opposed to having wrappers for the generic models that are implemented in `msprime.Demography`. @petrelharp do you have an opinion?

Author: I wasn't aware of this discussion. So it might be best not to discuss this so explicitly in the docs. We can simply mention that they exist, and if someone is interested in contributing an additional generic model, they should simply open an issue on GitHub to discuss this first. How does that sound?

Reviewer 2: Sounds fine!

.. _sec_development_demographic_model:

********************
Adding a new species
********************

---------------------------------------------------
Which information do I need to have for my species?
---------------------------------------------------

In `stdpopsim`, we aim to be inclusive and facilitate adding a diverse range of species.
That said, there are certain basic requirements we have
for every species added to the :ref:`sec_catalog`.
We specify these requirements below.
If you are unsure whether your species satisfies these baseline requirements,
but you still think it will be useful to add it to `stdpopsim`,
then we encourage you to open an `issue <http://github.com/popgensims/stdpopsim/issues>`_
on the GitHub repository to discuss this.
Others researchers in the community may be able to help you fill in the missing details
or find other solutions.

Every species added to `stdpopsim` should have the following information available:

1. A chromosome-level genome assembly
2. Mutation rate (per generation)
3. Recombination rate (per generation)
4. A characteristic population size
5. An average generation time

> 

Reviewer: ```suggestion

Of course, many species do not have precise estimates of each of these
(e.g., mutation rates are usually not known).
So, in practice we often substitute estimates from other species:
see below for more discussion.

```

Author: Will incorporate into my next commit.

Having a good quality assembly with complete chromosomes, or at least very long scaffolds,
is essential for chromosome-level simulations produced by `stdpopsim`.
Species with less complete genome builds typically do not have genetic maps
or good estimates of recombination rates,
making chromosome-level simulation much less useful.
Thus, currently, `stdpopsim` only supports adding species with near-complete
chromosome-level genome assemblies (i.e., close to one contig per chromosome).

An **average mutation rate** and **average recombination rate**
should be specified for each chromosome (per generation per bp).
The mutation rate estimate can be based on sequence data from pedigrees, mutation accumulation studies,
or comparative genomic analysis calibrated by fossil data (i.e., phylogenetic estimates).
Ideally, one would want to specify a fine-scale chromosome-level **recombination map**,
since the recombination rate is known to vary widely across chromosomes.
If a recombination map exists for your species,
you may specify it separately (see `Adding a genetic map`_).
Nonetheless, you should specify a default (average) recombination rate for each chromosome.

> If there is no information on the variation of rates among chromosome,

Reviewer: ```suggestion
If there is no information on the variation of rates along the chromosomes,
```

Author: I'm referring here to the case where you have an estimate only for the genome-wide average recombination rate and you wish to assign it to all chromosomes. So, variation **along** chromosomes doesn't seem to fit well. Maybe **across**, or **between**?

Author: I changed this to **variation of mutation rates across chromosomes** in my latest commit (also split mutation and recombination rates to separate sections here).

the average genome-wide mutation (or recombination) rate can be specified for all chromosomes.
Finally, if your species of interest does not have direct estimates of mutation (or recombination) rates,

> you may use estimates inferred for some closely related species.

Reviewer: ```suggestion
we recommend using estimates for some closely related species.
```

Author: Will incorporate in my next commit

The **effective population size** should represent the historical average effective population size,

> and produce simulated data that matches the average observed genetic diversity in that species.

Reviewer: ```suggestion
and produce simulated data that at least roughly matches the average observed genetic diversity in that species.
```

Author: This is the language we use in the manuscript. I think that the objective is to match the average genetic diversity. I understand that sometime people will choose another value that represents the field's consensus (like 10,000 for humans), even if it doesn't capture the observed average diversity. Still, I think that we should specify this as the objective of this number. How about adding **ideally** to this sentence:
> and ideally produce simulated data that matches the average . . .

Author: I went with *ideally produce simulated data that matches the average . . .*

Reviewer: I don't mind if you lleave it the way it is; but my point is that in practice it'll probably exactly match the diversity in whatever sample was used in some particular paper, but this will only roughly match future samples (and thus, "the species").

> This will not capture features of genetic variation that are caused by recent changes in population size and the presence of population structure.

Reviewer: ```suggestion
However, this will often not capture features of genetic variation
caused by recent changes in population size and the presence of population structure.
```

Author: Will incorporate this edit in my next commit

> To capture those, one should separately specify a fairly detailed demographic model for the species

Reviewer: ```suggestion
To capture those, one should also provide demographic model(s) for the species
```

Author: Will incorporate this edit in my next commit

(see `Adding a new demographic model`_).

> The **averge generation time** is an important part of the species' natural history,

Author: I'll apply this edit in my next commit

For example, the template files for *A. thaliana* were generated by executing this command:

.. code-block:: shell

    python -m maintenance add-species arabidopsis_thaliana

The `maintenance` utility generates three new files inside the `catalog/{species_id}/` directory:

* `__init__.py`
* `genome_data.py`
* `species.py`

The `__init__.py` file is a wrapper script that loads all the relevant information for your species. It should be edited only when you add components to your species, such as demographic models,
genetic maps, or DFEs.

The `genome_data.py` file contains the physical map of the genome.
The `maintenance` utility generates this file autmatically
based on information it downloaded from Ensembl.
Essentially, this utility sucks down a whole lot of useful information for free.
The `genome_data.py` file contains a data dictionary which has slots for the
assembly accession number, the assembly name,
and a dictionary representing the chromosome names and their associated lengths.
If synonyms are defined (e.g., chr2L for 2L) then those are given in the list that follows.
You should double-check the downloaded values, but there is probably no reason to edit this file.

> The `species.py` file

Reviewer: This looks like it should be a header of some kind

Author: Actually not. I didn't want to have too many short subsections. The sentence just got cut. I changed it to:
> The `species.py` file is where you specify the mutation and recombination rates, effective population size, and the average generation time, along with all accompanying citations (see details in **Which information do I need to have for my species?**).

Author: I restructured this a bit now. See latest commit.

(taken from the  `genome_data.py` file), as well as **mutation and recombination rates**, and **ploidy**.

For example, to define the `_genome` object for *A. thaliana*, we start by
defining auxiliary objects (`_recombination_rate_data` and `_ploidy`)
that specify the recombination rate and ploidy
for each chromosome.
Note that the mitochondrial and plastid genomes are associated
with ploidy of 1 and recombination rate of 0,
and all other chromosomes are associated with a ploidy of 2 and the
genome-wide average recombination rate.

.. code-block:: python

    _mean_recombination_rate = 200 / 124000 / 2 / 1e6
    _recombination_rate_data = {str(j): _mean_recombination_rate for j in range(1, 6)}
    _recombination_rate_data["Mt"] = 0
    _recombination_rate_data["Pt"] = 0

    _species_ploidy = 2
    _ploidy = {
      "1": _species_ploidy,
      "2": _species_ploidy,
      "3": _species_ploidy,
      "4": _species_ploidy,
      "5": _species_ploidy,
      "6": _species_ploidy,
      "Mt": 1,
      "Pt": 1,
    }

We then use another object (`_chromosomes`) to wrap up these
values together with the mutation rate, which is set to the
genome-wide average value for every chromosome:

> .. code-block:: python

Reviewer: There's a nicer way to do this, which is mostly what we use now: see [here](https://github.com/popsim-consortium/stdpopsim/blob/main/stdpopsim/catalog/GasAcu/species.py#L90) for instance, using [this function](https://github.com/popsim-consortium/stdpopsim/blob/main/stdpopsim/genomes.py#L45).

Author: Yes, this is cleaner. I can update the example to fit this cleaner form. Probably not in the next commit that I plan to push later tonight. I assume that I should modify the actual code in AraTha to fit what I use in the doc examples, right?

Author: I still need to do this. I want to take care of all the other doc issues first.

Author: @petrelharp I rewrote the AraTha example in the docs in my latest commit. Can you check this?

Reviewer 2: > I assume that I should modify the actual code in AraTha to fit what I use in the doc examples, right?

We can do this in a later PR (e.g. after this release) -- presumably, whomever is looking at the source code for guidance will be looking at the github head. Maybe just make an issue to point out the discrepancy.

Reviewer: Looks great! (I haven't tested it runs, but looks good?)

Adding a DFE model
******************

> 

Author: I modified the DFE section of the docs to mirror previous sections (adding species and demographic models). This probably should be reviewed by whomever wrote this initially, to make sure it's still accurate.

Reviewer 2: The rewrite looks fine to be. We can give it another look before selection paper/v0.2.1, too.

Submitting a PR and reviewing a DFE model
-----------------------------------------

After you finished your implementation, and specified all the
necessary citations,
you should submit a pull request (PR) with your changes to the code.
See the `GitHub workflow`_ for more details about this process.

At this point, most of your work is done.
**You have officially joined the** ``stdpopsim`` **development team. Welcome!!**
Your DFE model still needs to undergo review by another member
of the development team before it is fully incorporated into ``stdpopsim``.
This will likely require additional feedback from you,
so, stay tuned for discussion during the review process.

The review process for DFE models is similar to that of
new species or demographic models.
See the `Overview of the stdpopsim review process`_.
There is no template issue for a DFE model QC,
so the **QC issue** is generated by
`opening a new blank issue <https://github.com/popsim-consortium/stdpopsim/issues/new>`__
and specifying the following information in that issue:

1. **PR for new model:**
2. **Original paper:**
3. **Parameter values:**
4. **Potential issues:**
5. **QC'er requests:**

> 

Author: I think that this requires some more explanation. What should the contributor specify in the parameter values and potential issues. What do we mean by QCer requests? Does the contributor solicit specific people to review their PR?

We should also describe what the reviewer is expected to check. Is this just done manually, or is there any blind implementation involved here?

Reviewer 2: I agree, but we can also just make an issue to track the problem and leave off adding these details for now. The DFE QC stuff is still a bit up in the air (I think?), but doesn't really need to be clarified until v0.2.1 which will accompany the selection manuscript.

Author: Okay. We can keep this a bit vague for now and I'll open an issue so that we remember to add these details to the docs before the next release with all the selection stuff.
